### DISCUSSION TITLE
Why do you consider switching to Poetry?

### DISCUSSION THREAD

**Discussion by flying-sheep** [pydantic/pydantic#8043] [Question] [answered]
In https://github.com/pydantic/pydantic/pull/4473, you switched to hatch, which is a great move.

In https://github.com/python-poetry/poetry/issues/3332#issuecomment-1798718747, @samuelcolvin mentions

> we in the Pydantic team are keen to adopt poetry

I wonder why. Hatch is great, I use it for all projects, and its env mechanism makes testing and building docs a blast.

What does Poetry have that makes you consider switching?

**Comment by Viicos** [comment]
Poetry is great as an all in one tool, they probably want to be able to only use one instead of `pdm` + `hatch`.
Things need to be considered when using it for a library though: https://iscinumpy.dev/post/poetry-versions/

**Comment by samuelcolvin** [comment]
No particularly scientific reason, but something like:
* I find myself fighting with poetry a bit less than pdm
* poetry seems to install lots of dependencies slightly more quickly than pdm for me
* Some other members of the pydantic team prefer poetry

@adriangb as a fan of poetry, do you have anything more concrete/specific to add?

**Reply by adriangb** [reply, marked as answer]
In particular for a private monorepo use case we have we tried Pants, hatch + scripts, pdm and poetry, and poetry has been by far the least painful of all three. In my experience it has the most user friendly docs, errors and CLI.

**Reply by flying-sheep** [reply]
I see! If Hatch doesn’t have explicit support for this use case, that makes sense. @ofek might be interested in bringing it up to par regardless.

People are working on more narrative / use case driven docs for Hatch AFAIK.

**Reply by adriangb** [reply]
https://github.com/pypa/hatch/issues/233#issuecomment-1240920949

**Reply by adriangb** [reply]
Also fwiw hatch and pdm both tend to do the right thing (wait until there’s an official lock file format to support it and similar choices) while poetry does the “wrong” thing (like not support PEP-621) instead focusing on ease of use and UX.

**Reply by flying-sheep** [reply]
I usually build libraries, where a lockfile alone (without a dependency bumping bot) makes things harder, so the lack of one simplifies things for me. Library users wouldn’t use our lockfile, so having CI report the same incompatibilities that users run into allows us to fix them faster than waiting for users to report them.

But that’s beside the point: For applications, I know lock files make things easier.

**Reply by adriangb** [reply]
That was just an example. Another example is PDM's support for `__pypackages__` which in my experience has only gotten in the way. And that turned out to not even be the right thing because the PEP got rejected.
